src/wp-includes/rest-api/endpoints/class-wp-rest-block-directory-controller.php

```diff
@@ -0,0 +1,518 @@
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		$response = @plugins_api(
+			'query_plugins',
+			array(
+				'block'    => $search_string,
+				'per_page' => 3,
+			)
+		);
+
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$result = array();
+
+		foreach ( $response->plugins as $plugin ) {
+			$installed_plugins = get_plugins( '/' . $plugin['slug'] );
+
+			// Only show uninstalled blocks.
+			if ( empty( $installed_plugins ) ) {
+				$result[] = self::parse_block_metadata( $plugin );
+			}
+		}
+
+		return rest_ensure_response( $result );
+	}
+
+	/**
+	 * Determine if the endpoints are available.
+	 *
+	 * Only the 'Direct' filesystem transport, and SSH/FTP when credentials are stored are supported at present.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return bool|WP_Error True if filesystem is available, WP_Error otherwise.
+	 */
+	private static function is_filesystem_available() {
+		$filesystem_method = get_filesystem_method();
+
+		if ( 'direct' === $filesystem_method ) {
+			return true;
+		}
+
+		ob_start();
+		$filesystem_credentials_are_stored = request_filesystem_credentials( self_admin_url() );
+		ob_end_clean();
+
+		if ( $filesystem_credentials_are_stored ) {
+			return true;
+		}
+
+		return new WP_Error( 'fs_unavailable', __( 'The filesystem is currently unavailable for installing blocks.', 'gutenberg' ) );
+	}
+
+	/**
+	 * Parse block metadata for a block
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param WP_Object $plugin The plugin metadata.
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	private static function parse_block_metadata( $plugin ) {
+		$block = new stdClass();
+
+		// There might be multiple blocks in a plugin. Only the first block is mapped.
+		$block_data   = reset( $plugin['blocks'] );
+		$block->name  = $block_data['name'];
+		$block->title = $block_data['title'];
+		if ( ! $block->title ) {
+			$block->title = $plugin['name'];
+		}
+
+		// Plugin's description, not description in block.json.
+		$block->description = wp_trim_words( wp_strip_all_tags( $plugin['description'] ), 30, '...' );
+
+		$block->id                  = $plugin['slug'];
+		$block->rating              = $plugin['rating'] / 20;
+		$block->rating_count        = intval( $plugin['num_ratings'] );
+		$block->active_installs     = $plugin['active_installs'];
+		$block->author_block_rating = $plugin['author_block_rating'] / 20;
+		$block->author_block_count  = intval( $plugin['author_block_count'] );
+
+		// Plugin's author, not author in block.json.
+		$block->author = wp_strip_all_tags( $plugin['author'] );
+
+		// Plugin's icons or icon in block.json.
+		$block->icon = isset( $plugin['icons']['1x'] ) ? $plugin['icons']['1x'] : 'block-default';
+
+		$block->assets = array();
+		foreach ( $plugin['block_assets'] as $asset ) {
+			// TODO: Return from API, not client-set.
+			$block->assets[] = 'https://plugins.svn.wordpress.org/' . $plugin['slug'] . $asset;
```

Comment:
We should really look at returning that from the API side before this is merged into core